### DISCUSSION TITLE
Can I narrow required properties from a descendant?

### DISCUSSION THREAD

**Discussion by doomchild** [dotnet/csharplang#7972] [General]
Here's the sort of thing I'd like to be able to write:

```csharp
public record Base
{
  public required string BaseThing { get; init; }
}

public record Derived
{
  public required string DerivedThing { get; init; }

  public Derived()
  {
    BaseThing = "abc";
  }
}

Derived d = new()
{
  DerivedThing = "def"
}
```

But I can't do that, because the instantiation of `Derived` still wants `BaseThing` set in the object initializer. I know I can use `[SetsRequiredMembers]` on the constructors, but then I lose the compiler warning if `DerivedThing` isn't set in the object initializer.  Is there a way to effectively narrow the list of required fields in descendant types?  I vastly prefer using object initializers, so I was really looking forward to being able to use the `required` keyword, but when we updated to 8 (we only use LTS releases, so I didn't try this on 7), I ran into this problem.

**Comment by TahirAhmadov** [comment]
The official response from the LDT is to 1) decorate the ctor with `[SetsRequiredMembers]`, then 2) write your own analyzer which warns per your requirements.
People asked for `[SetsMember(nameof(BaseThing))]` kind of attribute but they rejected that idea for the language itself.

**Comment by doomchild** [comment]
Argh. I really like having the strictness of a compiler error, but I want to be able to specify base properties in descendant types (for things like defining more specific versions of general things). We do this now with records using primary constructors (where I _can_ specify values for base properties), but I vastly prefer object initializers to constructor calls wherever possible.

It sounds like I'm just SOL on this. How disappointing.

**Reply by TahirAhmadov** [reply]
You can promote your analyzer warnings to errors with a project or a Directory.Build.props file setting.
